gui: Do not translate InitWarning messages in debug.log

hebasto opened the pull request:
This PR forces the `bitcoin-qt` to write `InitWarning()` messages to the `debug.log` file in untranslated form, i.e., in English.

On master (376294cde6b1588cb17055d8fde567eaf5848c3c):
```
$ ./src/qt/bitcoin-qt -lang=nl -debug=vladidation -printtoconsole | grep 'vladi'
Warning: Niet-ondersteunde logcategorie -debug=vladidation.
2020-05-09T12:39:59Z Warning: Niet-ondersteunde logcategorie -debug=vladidation.
2020-05-09T12:40:02Z Command-line arg: debug="vladidation"
```

With this PR:
```
$ ./src/qt/bitcoin-qt -lang=nl -debug=vladidation -printtoconsole | grep 'vladi'
Warning: Unsupported logging category -debug=vladidation.
2020-05-09T12:42:04Z Warning: Unsupported logging category -debug=vladidation.
2020-05-09T12:42:35Z Command-line arg: debug="vladidation"
```

![Screenshot from 2020-05-09 15-42-31](https://user-images.githubusercontent.com/32963518/81474073-c7a50e00-920b-11ea-8775-c41122dacafe.png)

Related to #16218.

hebasto committed:
util: Enhance bilingual_str

hebasto committed:
util: Enhance Join()

hebasto committed:
gui: Do not translate InitWarning messages in debug.log

MarcoFalke commented on src/wallet/load.cpp:
```diff
@@ -56,7 +56,7 @@ bool VerifyWallets(interfaces::Chain& chain, const std::vector<std::string>& wal
         std::vector<bilingual_str> warnings;
         bool verify_success = CWallet::Verify(chain, location, salvage_wallet, error_string, warnings);
-        if (!warnings.empty()) chain.initWarning(Join(warnings, "\n", OpTranslated));
+        if (!warnings.empty()) chain.initWarning(Join(warnings, "\n", [](const bilingual_str& i) { return i; }));
```
```suggestion
        if (!warnings.empty()) chain.initWarning(Join(warnings, Untranslated("\n"));
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626197278).

MarcoFalke commented on src/util/translation.h:
```diff
@@ -16,13 +16,26 @@
+        return *this;
+    }
+
+    bilingual_str& operator+=(const std::string& chars)
```
I think it would be safer to not allow this and force the caller to go through `Untranslated`

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626197278).

MarcoFalke commented on src/util/string.h:
```diff
@@ -31,9 +31,10 @@ NODISCARD inline std::string TrimString(const std::string& str, const std::strin
  */
 template <typename T, typename UnaryOp>
-std::string Join(const std::vector<T>& list, const std::string& separator, UnaryOp unary_op)
+auto Join(const std::vector<T>& list, const std::string& separator, UnaryOp unary_op)
```
```suggestion
auto Join(const std::vector<T>& list, const U& separator, UnaryOp unary_op)
```

I think this requirement can be lifted (here and below)

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626197278).

maflcko reviewed (commented):
Didn't know that `auto` return type is possible in C++11, but it seems to compile so why not :man_shrugging:

hebasto commented:
Updated 4e789b9459ea3a5724cb30cccdaeb0d801f56e15 -> 6249e64acdcd865c44c74d311a079094f866fb75 ([pr18922.01](https://github.com/hebasto/bitcoin/commits/pr18922.01) -> [pr18922.02](https://github.com/hebasto/bitcoin/commits/pr18922.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18922.01..pr18922.02)):

- reworked following @MarcoFalke's ideas and suggestions

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2737,7 +2737,7 @@ static UniValue createwallet(const JSONRPCRequest& request)
     UniValue obj(UniValue::VOBJ);
     obj.pushKV("name", wallet->GetName());
-    obj.pushKV("warning", Join(warnings, "\n", OpOriginal));
+    obj.pushKV("warning", Join(warnings, std::string("\n"), OpOriginal));
```
Please keep the explicit overload for std::string, if you want you can extend it with a comment why it exists.

Also, I like the idea of deriving the return type from the unary operator

```diff
diff --git a/src/util/string.h b/src/util/string.h
index 81d4c186e5..e7651c3e7d 100644
--- a/src/util/string.h
+++ b/src/util/string.h
@@ -31,9 +31,10 @@ NODISCARD inline std::string TrimString(const std::string& str, const std::strin
  * @param unary_op   Apply this operator to each item in the list
  */
 template <typename T, typename BaseType, typename UnaryOp>
-BaseType Join(const std::vector<T>& list, const BaseType& separator, UnaryOp unary_op)
+auto Join(const std::vector<T>& list, const BaseType& separator, UnaryOp unary_op)
+    -> decltype(unary_op(list.at(0)))
 {
-    BaseType ret;
+    decltype(unary_op(list.at(0))) ret;
     for (size_t i = 0; i < list.size(); ++i) {
         if (i > 0) ret += separator;
         ret += unary_op(list.at(i));
@@ -47,6 +48,12 @@ T Join(const std::vector<T>& list, const T& separator)
     return Join(list, separator, [](const T& i) { return i; });
 }
 
+inline std::string Join(const std::vector<std::string>& list, const std::string& separator)
+{
+    // explicit overload needed for c_str, which would otherwise cause a substitution failure in the template above
+    return Join<std::string>(list, separator);
+}
+
 /**
  * Check if a string does not contain any embedded NUL (\0) characters
  */

hebasto replied:
> Please keep the explicit overload for std::string, if you want you can extend it with a comment why it exists.

Is it just to avoid writing `std::string("\n")` instead of `"\n"` in the `Join()` argument list?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626273821).

hebasto commented:
Updated 6249e64acdcd865c44c74d311a079094f866fb75 -> 3ddfa414787e017d9760b2ac55517986673a4b1c ([pr18922.02](https://github.com/hebasto/bitcoin/commits/pr18922.02) -> [pr18922.03](https://github.com/hebasto/bitcoin/commits/pr18922.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18922.02..pr18922.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18922#discussion_r422521418)

elichai commented:
> Didn't know that `auto` return type is possible in C++11, but it seems to compile so why not man_shrugging

That's interesting, I thought so too.
But apparently IIUC C++11 supports auto return type, but requires you to put a trailing return type (`-> return_type`) from C++14 you can also skip that and the return type will be deduced.

MarcoFalke commented on src/rpc/util.cpp:
```diff
@@ -751,7 +751,7 @@ std::string RPCArg::ToString(const bool oneline) const
     case Type::OBJ:
     case Type::OBJ_USER_KEYS: {
-        const std::string res = Join(m_inner, ",", [&](const RPCArg& i) { return i.ToStringObj(oneline); });
+        const std::string res = Join(m_inner, std::string(","), [&](const RPCArg& i) { return i.ToStringObj(oneline); });
```
Why are all these unrelated changes needed. This diff compiles fine for me on top:


```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 597da6e7fb..860fa198d5 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -751,7 +751,7 @@ std::string RPCArg::ToString(const bool oneline) const
     }
     case Type::OBJ:
     case Type::OBJ_USER_KEYS: {
-        const std::string res = Join(m_inner, std::string(","), [&](const RPCArg& i) { return i.ToStringObj(oneline); });
+        const std::string res = Join(m_inner, ",", [&](const RPCArg& i) { return i.ToStringObj(oneline); });
         if (m_type == Type::OBJ) {
             return "{" + res + "}";
         } else {
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
index f71e7efa2b..45b7fd4932 100644
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -141,15 +141,15 @@ BOOST_AUTO_TEST_CASE(util_HexStr)
 BOOST_AUTO_TEST_CASE(util_Join)
 {
     // Normal version
-    BOOST_CHECK_EQUAL(Join({}, std::string(", ")), "");
-    BOOST_CHECK_EQUAL(Join({"foo"}, std::string(", ")), "foo");
-    BOOST_CHECK_EQUAL(Join({"foo", "bar"}, std::string(", ")), "foo, bar");
+    BOOST_CHECK_EQUAL(Join({}, ", "), "");
+    BOOST_CHECK_EQUAL(Join({"foo"}, ", "), "foo");
+    BOOST_CHECK_EQUAL(Join({"foo", "bar"}, ", "), "foo, bar");
 
     // Version with unary operator
     const auto op_upper = [](const std::string& s) { return ToUpper(s); };
-    BOOST_CHECK_EQUAL(Join<std::string>({}, std::string(", "), op_upper), "");
-    BOOST_CHECK_EQUAL(Join<std::string>({"foo"}, std::string(", "), op_upper), "FOO");
-    BOOST_CHECK_EQUAL(Join<std::string>({"foo", "bar"}, std::string(", "), op_upper), "FOO, BAR");
+    BOOST_CHECK_EQUAL(Join<std::string>({}, ", ", op_upper), "");
+    BOOST_CHECK_EQUAL(Join<std::string>({"foo"}, ", ", op_upper), "FOO");
+    BOOST_CHECK_EQUAL(Join<std::string>({"foo", "bar"}, ", ", op_upper), "FOO, BAR");
 }
 
 BOOST_AUTO_TEST_CASE(util_FormatParseISO8601DateTime)
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index 0534befd3d..dda00f1fe7 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -2597,7 +2597,7 @@ static UniValue loadwallet(const JSONRPCRequest& request)
 
     UniValue obj(UniValue::VOBJ);
     obj.pushKV("name", wallet->GetName());
-    obj.pushKV("warning", Join(warnings, std::string("\n"), OpOriginal));
+    obj.pushKV("warning", Join(warnings, "\n", OpOriginal));
 
     return obj;
 }
@@ -2737,7 +2737,7 @@ static UniValue createwallet(const JSONRPCRequest& request)
 
     UniValue obj(UniValue::VOBJ);
     obj.pushKV("name", wallet->GetName());
-    obj.pushKV("warning", Join(warnings, std::string("\n"), OpOriginal));
+    obj.pushKV("warning", Join(warnings, "\n", OpOriginal));
 
     return obj;
 }

hebasto replied:
You are right. Those are remnants from the previous version of this branch.

hebasto replied:
[Fixed](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626331367).

hebasto commented:
Updated 3ddfa414787e017d9760b2ac55517986673a4b1c -> 48710740c6f1cfa37c20545d0d6179d5de4bae6e ([pr18922.03](https://github.com/hebasto/bitcoin/commits/pr18922.03) -> [pr18922.04](https://github.com/hebasto/bitcoin/commits/pr18922.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18922.03..pr18922.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18922#discussion_r422644375)

MarcoFalke commented on src/logging.h:
```diff
@@ -137,7 +137,7 @@ namespace BCLog {
         std::string LogCategoriesString()
         {
-            return Join(LogCategoriesList(), ", ", [&](const LogCategory& i) { return i.category; });
+            return Join(LogCategoriesList(), std::string(", "), [&](const LogCategory& i) { return i.category; });
```
Why is this one needed? (Sorry I missed it in my last review comment)

hebasto replied:
Thank you. [Fixed](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626343159).

hebasto commented:
Updated 48710740c6f1cfa37c20545d0d6179d5de4bae6e -> da16f95c3fecf4ee1e9a1dc4333b0b92cd981afd ([pr18922.04](https://github.com/hebasto/bitcoin/commits/pr18922.04) -> [pr18922.05](https://github.com/hebasto/bitcoin/commits/pr18922.05), [diff](https://github.com/hebasto/bitcoin/compare/pr18922.04..pr18922.05)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18922#discussion_r422654602):
> Why is this one needed? (Sorry I missed it in my last review comment)

---

The current implementation of the `Join()` allows do not use `OpOriginal` and `OpTranslated` unary operators at all.
Is it worth to drop them?

MarcoFalke commented:
>  Is it worth to drop them?

Up to you. I am happy to review either approach, as long as this is done in a new commit.

hebasto committed:
util: Drop OpOriginal() and OpTranslated()

The current implementation of the Join() allows do not use OpOriginal()
and OpTranslated() unary operators at all.

hebasto commented:
Added commit which removes `OpOriginal()` and `OpTranslated()` unary operators.

promag reviewed (commented):
Concept ACK, code looks fine.

laanwj commented:
vladiation lol :joy: 

Looks good to me. Nice cleanup with the `*Op` functions.

ACK 78be8d97d3d750fcfbbe509a72639b7b30b7bd18

To have an idea of progress, I would like to know whether an end to this in in sight. Is `InitWarning` the last step? Or are there more functions after this that need to be converted to `bilingual_str`?

jonasschnelli commented:
Have vladidated the code.

utACK 78be8d97d3d750fcfbbe509a72639b7b30b7bd18

MarcoFalke commented:
I think the last ones are AbortNode and the gui/rpc warnings https://github.com/bitcoin/bitcoin/issues/16218#issuecomment-626320627

MarcoFalke commented:
ACK 78be8d97d3d750fcfbbe509a72639b7b30b7bd18 📢

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 78be8d97d3d750fcfbbe509a72639b7b30b7bd18 📢
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUghsAv/TBQtsX9Zg+DSmubigFhi4U+8zCCTcrY+wH1oBCBcRVZdyDrgKQzWSQH5
f6fg99QfwzQYdhQd3Yykh1+YYTd0uHiUX/+YgRlaXpqNqvMoijMX/JjFgEOOKOCD
rueYNroFEqBChN79zQXj01/SIrVifVtIHzHQTbPsBkcQ6vUInOkQ9eBOpJZvkxro
6WTC/APvUWZWS1F8prsxOR/t6a1YVg8RdjpKqfnQIxUYj/uPIVQfDYK5nhyo4Tao
FRmYA6pDoDsT3MYkDFDOCG/4dXyri8PTdVlqv2W0fyJ5fjGT1YGIyrMp56aILRlj
CtaIniMEj+gQIP2Aqn49hCzlYP7ljMg+IP5BVKBbPu4C0fCjf1QFPYyDMrUT2yP0
w1/7NXqEerOFDpTewkO0C/a96sGN/TpvjDr4uaZ4Svx32kF0pmhiLxy0qtjwL7Wg
6iGt+m+yHDGPsJWzxDAlm+sJXzlUd3s2sQpDlqwv5afHASC5re38Vhc6OaqGt3lz
WrKEK80Y
=Eo5p
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `69702236ca2cac943236ee37ffe813ced8a7bfccdd4b1792664d9fe1671d2185  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010869702236ca2cac943236ee37ffe813ced8a7bfccdd4b1792664d9fe1671d2185f010128dbf8ef86a4710b6348943522851e608fff0100800ba285670b18fc00f2dab2b398f4c08f1045ebc3b7ff008d9b00808833119de0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01013f975039b0bd08acaf561193ad4600608f1045ebc3b7ff00871acf815af739f660083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0102238dee13464af93267992773c4e446808f1045ebc3b7ff00835df154f85c0ed590083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01078e52bbf8396f39a0d3caa80bc6969b908f1045ebc3b80f008d5503b231c1a07b60083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
@jonasschnelli Ready for merge?!
